Healpix doc (#2498)
nasa-gcn/gcn.nasa.gov: app/routes/docs.client.samples.md

[MOCpy](https://cds-astro.github.io/mocpy/): Python library allowing easy creation, parsing and manipulation of Multi-Order Coverage maps.

## Papers

(1) Calabretta, M. R., & Roukema, B. F. 2007, Mon. Notices Royal Astron. Soc., 381, 865. [doi:10.1111/j.1365-2966.2007.12297.x](https://academic.oup.com/mnras/article/381/2/865/1021805?login=true)

(2) Górski, K.M., Hivon, E., Banday, A.J., et al. 2005, Astrophys. J., 622, 759. [doi:10.1086/427976](https://iopscience.iop.org/article/10.1086/427976)

(3) Górski, K. M., Wandelt, B. D., et al. 1999. [arXiv:astro-ph/9905275](https://arxiv.org/abs/astro-ph/9905275)

(4) Fernique, P., Allen, et al. 2015, Astron. Astrophys., 578, A114. [doi:10.1051/0004-6361/201526075](https://www.aanda.org/articles/aa/full_html/2015/06/aa26075-15/aa26075-15.html)

(5) Fernique, P., Boch, T., et al. 2014, IVOA Recommendation. [arXiv:1505.02937](https://arxiv.org/abs/1505.02937)

(6) Martinez-Castellanos, I., Singer, L. P., et al. 2022, Astrophys. J., 163, 259. [doi:10.3847/1538-3881/ac6260](https://iopscience.iop.org/article/10.3847/1538-3881/ac6260)

> (7) Singer, L. P., & Price, L. R. 2016, Phys. Rev. D, 93, 024013. [doi:10.1103/PhysRevD.93.024013](https://journals.aps.org/prd/abstract/10.1103/PhysRevD.93.024013)

Reviewer: Please use DOIs for all paper URLs.

Reviewer 2: resolved in 569e18c

            continue
    print(message.value())
```

## HEALPix Sky Maps

[HEALPix](https://healpix.sourceforge.io) (<b>H</b>ierarchical, <b>E</b>qual <b>A</b>rea, and iso-<b>L</b>atitude <b>Pix</b>elisation) is a scheme for indexing positions on the unit sphere.

> For localization of events, the multi-messenger community uses the standard HEALPix, @2005ApJ...622..759G with the file extension `.fits.gz`, as well as multi-resolution HEALPix, @2015A&A...578A.114F with the file extension `.multiorder.fits`. The preferred format is the multi-resolution HEALPix format.

Reviewer: ```suggestion
For localization of events, the multi-messenger community uses the standard HEALPix [@2005ApJ...622..759G] with the file extension `.fits.gz`, as well as multi-resolution HEALPix [@2015A&A...578A.114F] with the file extension `.multiorder.fits`. The preferred format is the multi-resolution HEALPix format.
```

### Multi-Order Sky Maps

GCN strongly encourages the use of multi-order sky maps. They utilize a variable resolution, with higher probability regions having higher resolution and lower probability regions being encoded with a lower resolution. Variable resolution allows multi-order sky maps to be significantly more efficient than single-resolution HEALPix sky maps with respect to both storage footprint and read speed.

#### Reading Sky Maps

> Sky maps can be parsed using Python and a small number of packages. While this documentation covers the use of `astropy-healpix`, there are several packages that can be used for this purpose; a number of [alternatives](#references) are listed at the bottom of this page.

Reviewer: The `#references` link is broken.

Author: Fixed!

hpx[ipix_max] / hp.nside2pixarea(nside, degrees=True)

# Highest probability pixel on the sky
ra, dec = hp.pix2ang(nside, ipix_max, lonlat=True)
ra, dec
```

#### Integrated probability in a Circle

We can calculate the integrated probability within an arbitrary circle on the sky

```python
# First define the Cartesian coordinates of the center of the circle
ra = 180.0
dec = -45.0
radius = 2.5

# Calculate Cartesian coordinates of the center of the Circle
xyz = hp.ang2vec(ra, dec, lonlat=True)

# Obtain an array of indices for the pixels inside the circle
ipix_disc = hp.query_disc(nside, xyz, np.deg2rad(radius))

# Sum the probability in all of the matching pixels:
hpx[ipix_disc].sum()
```

#### Integrated probability in a Polygon

Similar to the integrated probability within a circle, it is possible to calculate the integrated probability of the source lying within an arbitrary polygon

```python
#  Indices of the pixels within a polygon (defined by the Cartesian coordinates of its vertices)

xyz = [[0, 0, 0],
       [1, 0, 0],
       [1, 1, 0],
       [0, 1, 0]]
ipix_poly = hp.query_polygon(nside, xyz)
hpx[ipix_poly].sum()
```

#### Other Documentation and HEALPix Packages

> 

Author: ```suggestion
For more information and resources on the analysis and manipulation of pixelated data on a sphere, explore the following links:
```

> For more information and resources on the analysis and manipulation of pixelated data on a sphere, explore the following links:

Author: ```suggestion
For more information and resources on the analysis and manipulation of pixelated data on a sphere, explore the following links:

```

> For more information and resources on the analysis and manipulation of pixelated data on a sphere, explore the following links:

Author: ```suggestion

For more information and resources on the analysis and manipulation of pixelated data on a sphere, explore the following links:

```

Additional information can be found on the [LIGO website](https://emfollow.docs.ligo.org/userguide/tutorial/multiorder_skymaps.html)

[healpy](https://healpy.readthedocs.io/en/latest/): Official python library for handling the pixlated data on sphere

[astropy-healpix](https://pypi.org/project/astropy-healpix/): Integrates HEALPix with Astropy for data manipulation and analysis

> [mhealpy](https://mhealpy.readthedocs.io/en/latest/): Object-oriented wrapper of healpy for handling the multi-resolution maps

Reviewer: None of these are complete sentences. Should this be a bulleted list?

Author: Yup, agree

For localization of events, the multi-messenger community uses the standard HEALPix [@2005ApJ...622..759G] with the file extension `.fits.gz`, as well as multi-resolution HEALPix [@2015A&A...578A.114F] with the file extension `.multiorder.fits`. The preferred format is the multi-resolution HEALPix format.

### Multi-Order Sky Maps

GCN strongly encourages the use of multi-order sky maps. They utilize a variable resolution, with higher probability regions having higher resolution and lower probability regions being encoded with a lower resolution. Variable resolution allows multi-order sky maps to be significantly more efficient than single-resolution HEALPix sky maps with respect to both storage footprint and read speed.

#### Reading Sky Maps

Sky maps can be parsed using Python and a small number of packages. While this documentation covers the use of `astropy-healpix`, there are several packages that can be used for this purpose; a number of [alternatives](#other-documentation-and-healpix-packages) are listed at the bottom of this page.

```python
import astropy_healpix as ah
import numpy as np

from astropy import units as u
from astropy.table import QTable
```

> A given sky map can then be read in as

Reviewer: ```suggestion
A given sky map can then be read in as:
```

```python
skymap = QTable.read('skymap.multiorder.fits')
```

#### Most Probable Sky Location

> The index of the highest probability point

Reviewer: This is not a complete sentence.

> The index of the highest probability point

Author: ```suggestion
Let's find the index of the point in the sky with the highest probability density:
```

```python
hp_index = np.argmax(skymap['PROBDENSITY'])
uniq = skymap[hp_index]['UNIQ']

level, ipix = ah.uniq_to_level_ipix(uniq)
nside = ah.level_to_nside(level)

ra, dec = ah.healpix_to_lonlat(ipix, nside, order='nested')
```

#### Probability Density at a Known Position

> Similarly, one can calculate the probability density at a known position

Reviewer: ```suggestion
Similarly, one can calculate the probability density at a known position:
```

```python
ra, dec = 197.450341598 * u.deg, -23.3814675445 * u.deg

level, ipix = ah.uniq_to_level_ipix(skymap['UNIQ'])
nside = ah.level_to_nside(level)

match_ipix = ah.lonlat_to_healpix(ra, dec, nside, order='nested')

match_index = np.flatnonzero(ipix == match_ipix)[0]

prob_density = skymap[match_index]['PROBDENSITY'].to_value(u.deg**-2)
```

#### 90% Probability Region

The estimation of a 90% probability region involves sorting the pixels, calculating the area of each pixel, and then summing the probability of each pixel until 90% is reached.

```python
#Sort the pixels by decending probability density
skymap.sort('PROBDENSITY', reverse=True)

#Area of each pixel
level, ipix = ah.uniq_to_level_ipix(skymap['UNIQ'])
pixel_area = ah.nside_to_pixel_area(ah.level_to_nside(level))

#Pixel area times the probability
prob = pixel_area * skymap['PROBDENSITY']

#Cummulative sum of probability
cumprob = np.cumsum(prob)

#Pixels for which cummulative is 0.9
i = cumprob.searchsorted(0.9)

#Sum of the areas of the pixels up to that one
area_90 = pixel_area[:i].sum()
area_90.to_value(u.deg**2)
```

### Flat Resolution HEALPix Sky maps

> A sky map `.fits.gz` file can be read in using `healpy`

Reviewer: ```suggestion
A sky map `.fits.gz` file can be read in using `healpy`:
```

```python
import healpy as hp
import numpy as np
from matplotlib import pyplot as plt

# Read both the HEALPix image data and the FITS header
hpx, header = hp.read_map('skymap.fits.gz', h=True)

# Plot a Mollweide-projection all-sky image
np.mollview(hpx)
plt.show()
```

#### Most Probable Sky Location

> The point on the sky with the highest probability can be found by finding the maximum value in the HEALPix sky map

Reviewer: ```suggestion
The point on the sky with the highest probability can be found by finding the maximum value in the HEALPix sky map:
```

```python
# Reading Sky Maps with Healpy
healpix_image = hp.read_map('bayestar.fits.gz,0')
npix = len(hpx)

# Lateral resolution of the HEALPix map
nside = hp.npix2nside(npix)

# Find the highest probability pixel
ipix_max = np.argmax(hpx)

# Probability density per square degree at that position
hpx[ipix_max] / hp.nside2pixarea(nside, degrees=True)

# Highest probability pixel on the sky
ra, dec = hp.pix2ang(nside, ipix_max, lonlat=True)
ra, dec
```

#### Integrated probability in a Circle

> We can calculate the integrated probability within an arbitrary circle on the sky

Reviewer: ```suggestion
We can calculate the integrated probability within an arbitrary circle on the sky:
```

```python
# First define the Cartesian coordinates of the center of the circle
ra = 180.0
dec = -45.0
radius = 2.5

# Calculate Cartesian coordinates of the center of the Circle
xyz = hp.ang2vec(ra, dec, lonlat=True)

# Obtain an array of indices for the pixels inside the circle
ipix_disc = hp.query_disc(nside, xyz, np.deg2rad(radius))

# Sum the probability in all of the matching pixels:
hpx[ipix_disc].sum()
```

#### Integrated probability in a Polygon

> Similar to the integrated probability within a circle, it is possible to calculate the integrated probability of the source lying within an arbitrary polygon

Reviewer: ```suggestion
Similar to the integrated probability within a circle, it is possible to calculate the integrated probability of the source lying within an arbitrary polygon:
```

```python
#  Indices of the pixels within a polygon (defined by the Cartesian coordinates of its vertices)

xyz = [[0, 0, 0],
       [1, 0, 0],
       [1, 1, 0],
       [0, 1, 0]]
ipix_poly = hp.query_polygon(nside, xyz)
hpx[ipix_poly].sum()
```

#### Other Documentation and HEALPix Packages

For more information and resources on the analysis of pixelated data on a sphere, explore the following links:

- Additional information can be found on the [LIGO website](https://emfollow.docs.ligo.org/userguide/tutorial/multiorder_skymaps.html)

> - [healpy](https://healpy.readthedocs.io/en/latest/): Official python library for handling the pixlated data on sphere

Reviewer: ```suggestion
- [healpy](https://healpy.readthedocs.io/en/latest/): Official Python library for handling the pixelated data on sphere
```

            continue
    print(message.value())
```

## HEALPix Sky Maps

[HEALPix](https://healpix.sourceforge.io) (<b>H</b>ierarchical, <b>E</b>qual <b>A</b>rea, and iso-<b>L</b>atitude <b>Pix</b>elisation) is a scheme for indexing positions on the unit sphere.
For localization of events, the multi-messenger community uses the standard [HEALPix](https://healpix.sourceforge.io) format with the file extension `.fits.gz`, as well as multi-resolution HEALPix format with the file extension `.multiorder.fits`. The preferred format is the multi-resolution HEALPix format.

### Multi-Order Sky Maps

> GCN strongly encourages the use of multi-order sky maps. They utilize a variable resolution, with higher probability regions having higher resolution and lower probability regions being encoded with a lower resolution. This is signicantly more efficient than single-resolution HEALPix sky maps with respect to storage footprint and read speed. However, interpreting these multi-order sky maps can be more complex.

Reviewer 2: Typo for "significantly"

```python
ra, dec = 197.450341598 * u.deg, -23.3814675445 * u.deg

level, ipix = ah.uniq_to_level_ipix(skymap['UNIQ'])
nside = ah.level_to_nside(level)

match_ipix = ah.lonlat_to_healpix(ra, dec, nside, order='nested')

match_index = np.flatnonzero(ipix == match_ipix)[0]

prob_density = skymap[match_index]['PROBDENSITY'].to_value(u.deg**-2)
```

#### 90% Probability Region

The estimation of a 90% probability region involves sorting the pixels, calculating the area of each pixel, and then summing the probability of each pixel until 90% is reached.

```python
#Sort the pixels by decending probability density
skymap.sort('PROBDENSITY', reverse=True)

#Area of each pixel
level, ipix = ah.uniq_to_level_ipix(skymap['UNIQ'])
pixel_area = ah.nside_to_pixel_area(ah.level_to_nside(level))

#Pixel area times the probability
prob = pixel_area * skymap['PROBDENSITY']

#Cummulative sum of probability
cumprob = np.cumsum(prob)

#Pixels for which cummulative is 0.9
i = cumprob.searchsorted(0.9)

#Sum of the areas of the pixels up to that one
area_90 = pixel_area[:i].sum()
area_90.to_value(u.deg**2)
```

### Flat Resolution HEALPix Sky maps

> Let's say you have a sky map fits file. To read in Python with Healpy:

Author: Earlier looks fine

            continue
    print(message.value())
```

## HEALPix Sky Maps

> HEALPix stands for <b>H</b>ierarchical, <b>E</b>qual <b>A</b>rea, and iso-<b>L</b>atitude <b>Pix</b>elisation is a scheme for indexing positions on the unit sphere.

Reviewer 3: ```suggestion
HEALPix (<b>H</b>ierarchical, <b>E</b>qual <b>A</b>rea, and iso-<b>L</b>atitude <b>Pix</b>elisation) is a scheme for indexing positions on the unit sphere.
```

> For localization of events, the multi-messenger community uses the standard [HEALPix](https://healpix.sourceforge.io) format with the file extension `.fits.gz`, as well as multi-resolution HEALPix format with the file extension `.multiorder.fits`. The preferred format is the multi-resolution HEALPix format, that is only format explicitly listed in the GCN.

Reviewer 3: '.gz' isn't specific to a healpix file, it's just showing that it's zipped.
Also, what is meant by "listed in the GCN"?

Reviewer: No, it's not specific to HEALPix, but the convention for localization FITS files is that it's either a flat-resolution file and gzip-compressed, or a multi-order file and _not_ compressed.

### Multi-Order Sky Maps

> GCN strongly encourages the use of multi-order sky maps. These are a more complex form of healpix sky maps that utilize a variable resolution, with higher probability regions having higher resolution and lower probability regions being encoded with a lower resolution. This is signicantly more efficient than single-resolution healpix sky maps with respect to storage footprint and read speed. However, interpreting these multi-order sky maps can be more complex.

Reviewer 3: ```suggestion
GCN strongly encourages the use of multi-order sky maps. They utilize a variable resolution, with higher probability regions having higher resolution and lower probability regions being encoded with a lower resolution. This is signicantly more efficient than single-resolution HEALPix sky maps with respect to storage footprint and read speed. However, interpreting these multi-order sky maps can be more complex.
```

### Working with HEALPix Maps

> HEALPix data is distrubuted in standard format with the file extension `.fits.gz`. These files are FITS image files and can be utilized with many common FITS tools. These files usually stored as HEALPix projection,

Reviewer 3: incomplete sentence

> HEALPix data is distrubuted in standard format with the file extension `.fits.gz`. These files are FITS image files and can be utilized with many common FITS tools. These files usually stored as HEALPix projection,

Author: ```suggestion
HEALPix data is distrubuted in standard format with the file extension `.fits.gz`. These files are FITS image files and can be utilized with many common FITS tools. Typically, the data is stored using the HEALPix projection.
```

> HEALPix data is distrubuted in standard format with the file extension `.fits.gz`. These files are FITS image files and can be utilized with many common FITS tools. Typically, the data is stored using the HEALPix projection.

Reviewer 3: Is this new info compared to what's said in "## HEALPix Sky Maps" above?

#### Reading Sky Maps

> Sky maps can be parsed using python; to start, import a handful of packages (note: while this documentation covers the use of `astropy-healpix`, there are several packages that can be used for this purpose; a number of alternatives are listed at the bottom of this page)

Reviewer 3: You don't need to explain that you need to import packages.  Also, link to the alternatives at the bottom of the page.

Reviewer: No, I think that you really do need to explain that.

```python
import astropy_healpix as ah
import numpy as np

from astropy import units as u
from astropy.table import QTable
```

A given sky map can then be read in as

```python
skymap = QTable.read('skymap.multiofits)
```

#### Most Probable Sky Location

> The index of the highest probability point can be found by doing the following

Reviewer 3: ```suggestion
The index of the highest probability point
```

```python
hp_index = np.argmax(skymap['PROBDENSITY'])
uniq = skymap[hq_index]['UNIQ']

level, ipix = ah.uniq_to_level_ipix(uniq)
nside = ah.level_to_nside(level)

ra, dec = ah.healpix_to_lonlat(ipix, nside, order='nested')
```

#### Probability Density at a Known Position

Similarly, one can calculate the probability density at a known position

```python
ra, dec = 197.450341598 * u.deg, -23.3814675445 * u.deg

level, ipix = ah.uniq_to_level_ipix(skymap['UNIQ'])
nside = ah.level_to_nside(level)

match_ipix = ad.lonlat_to_healpix(ra, dec, nside, order='nested')

match_index = np.flatnonzero(ipix == match_ipix)[0]

prob_density = skymap[match_index]['PROBDENSITY'].to_value(u.deg**-2)
```

#### Find the 90% Probability Region

> The estimation of a 90% probability region, involves sorting the pixel, calculating the area of each pixel, and then proceeding by the probability of each pixel.

Reviewer 3: ```suggestion
The estimation of a 90% probability region, involves sorting the pixel, calculating the area of each pixel, sorting by probability, and then summing the probability of each pixel until 90% is reached.
```

Reviewer 3: Some of these examples explain how the calculation works, others just say you this is how you do it.  Either way works, but be consistent.

ra, dec
```

#### Integrated probability in a Circle

We call hp.query_disc to obtain an array of indices for the pixels inside the circle.

```python
# First define the Cartesian coordinates of the center of the Circle
ra = 180.0
dec = -45.0
radius = 2.5

# We convert to spherical polar coordinates
theta = 0.5 * np.pi - np.deg2rad(dec)
phi = np.deg2rad(ra)
radius = np.deg2rad(radius)

# Calculate Cartesian coordinates of the center of the Circle
xyz = hp.ang2vec(theta, phi)

# Obtain an array of indices for the pixels inside the circle
ipix_disc = hp.query_disc(nside, xyz, radius)

# Sum the probability in all of the matching pixels:
hpx[ipix_disc].sum()

```

#### Integrated probability in a Polygon

We can use the `hp.query_polygon` function to find the pixels inside a polygon and compute the probability that the source is inside the polygon by adding up the pixel values.

```python
#  Indices of the pixels within a polygon (defined by the Cartesian coordinates of its vertices)

xyz = [[-0.5, -0.4, -0.5],
        [-0.4, -0.4, -0.6],
        [-0.6, -0.3, -0.6],
        [-0.7 , -0.4, -0.5]]
ipix_poly = hp.query_polygon(nside, xyz)
hpx[ipix_poly].sum()
```

> ##### Further documentation

Reviewer 3: ```suggestion
##### Other Documentation and HEALPix Packages
```

Additional information can be found on the [LIGO website](https://emfollow.docs.ligo.org/userguide/tutorial/multiorder_sky maps.html)

[healpy](https://healpy.readthedocs.io/en/latest/): Official python library for handling the pixlated data on sphere

> [astropy-healpix](https://pypi.org/project/astropy-healpix/)

Reviewer 3: add description

> [astropy-healpix](https://pypi.org/project/astropy-healpix/)

Author: ```suggestion
[astropy-healpix](https://pypi.org/project/astropy-healpix/): Integrates HEALPix with Astropy for data manipulation and analysis
```

[mhealpy](https://mhealpy.readthedocs.io/en/latest/): Object-oriented wrapper of healpy for handling the multi-resolution maps

[MOCpy](https://cds-astro.github.io/mocpy/): Python library allowing easy creation, parsing and manipulation of Multi-Order Coverage maps.

> ## Papers

Reviewer 3: ```suggestion
## References
```

            continue
    print(message.value())
```

## HEALPix Sky Maps

HEALPix (<b>H</b>ierarchical, <b>E</b>qual <b>A</b>rea, and iso-<b>L</b>atitude <b>Pix</b>elisation) is a scheme for indexing positions on the unit sphere.

> For localization of events, the multi-messenger community uses the standard [HEALPix](https://healpix.sourceforge.io) format with the file extension `.fits.gz`, as well as multi-resolution HEALPix format with the file extension `.multiorder.fits`. The preferred format is the multi-resolution HEALPix format, that is only format explicitly listed in the GCN.

Reviewer: ```suggestion
For localization of events, the multi-messenger community uses the standard [HEALPix](https://healpix.sourceforge.io) format with the file extension `.fits.gz`, as well as multi-resolution HEALPix format with the file extension `.multiorder.fits`. The preferred format is the multi-resolution HEALPix format.
```

> For localization of events, the multi-messenger community uses the standard [HEALPix](https://healpix.sourceforge.io) format with the file extension `.fits.gz`, as well as multi-resolution HEALPix format with the file extension `.multiorder.fits`. The preferred format is the multi-resolution HEALPix format, that is only format explicitly listed in the GCN.

Reviewer: ```suggestion
For localization of events, the multi-messenger community uses the standard [HEALPix-in-FITS](https://healpix.sourceforge.io/data/examples/healpix_fits_specs.pdf) format with the file extension `.fits.gz`, as well as multi-resolution HEALPix format with the file extension `.multiorder.fits`. The preferred format is the multi-resolution HEALPix format. This is generally the format that is used in GCN Notices.
```

### Multi-Order Sky Maps

GCN strongly encourages the use of multi-order sky maps. They utilize a variable resolution, with higher probability regions having higher resolution and lower probability regions being encoded with a lower resolution. This is signicantly more efficient than single-resolution HEALPix sky maps with respect to storage footprint and read speed. However, interpreting these multi-order sky maps can be more complex.

#### Reading Sky Maps

> Sky maps can be parsed using python; to start, import a handful of packages (note: while this documentation covers the use of `astropy-healpix`, there are several packages that can be used for this purpose; a number of [alternatives](#references) are listed at the bottom of this page)

Reviewer: You also have healpy code samples here.

Reviewer: ```suggestion
Sky maps can be parsed using Python; to start, import a handful of packages (note: while this documentation covers the use of `astropy-healpix`, there are several packages that can be used for this purpose; a number of [alternatives](#references) are listed at the bottom of this page)
```

```python
import astropy_healpix as ah
import numpy as np

from astropy import units as u
from astropy.table import QTable
```

A given sky map can then be read in as

```python

> skymap = QTable.read('skymap.multiofits)

Reviewer: ```suggestion
skymap = QTable.read('skymap.multiorder.fits')
```

```

#### Most Probable Sky Location

The index of the highest probability point

```python
hp_index = np.argmax(skymap['PROBDENSITY'])

> uniq = skymap[hq_index]['UNIQ']

Reviewer: ```suggestion
uniq = skymap[hp_index]['UNIQ']
```

level, ipix = ah.uniq_to_level_ipix(uniq)
nside = ah.level_to_nside(level)

ra, dec = ah.healpix_to_lonlat(ipix, nside, order='nested')
```

#### Probability Density at a Known Position

> Similarly, one can calculate the probability density at a known position

Reviewer: ```suggestion
Similarly, one can calculate the probability density at a known position:
```

```python
ra, dec = 197.450341598 * u.deg, -23.3814675445 * u.deg

level, ipix = ah.uniq_to_level_ipix(skymap['UNIQ'])
nside = ah.level_to_nside(level)

> match_ipix = ad.lonlat_to_healpix(ra, dec, nside, order='nested')

Reviewer: ```suggestion
match_ipix = ah.lonlat_to_healpix(ra, dec, nside, order='nested')
```

match_index = np.flatnonzero(ipix == match_ipix)[0]

prob_density = skymap[match_index]['PROBDENSITY'].to_value(u.deg**-2)
```

#### 90% Probability Region

> The estimation of a 90% probability region, involves sorting the pixel, calculating the area of each pixel, sorting by probability, and then summing the probability of each pixel until 90% is reached.

Reviewer: ```suggestion
The estimation of a 90% probability region involves sorting the pixels, calculating the area of each pixel, and then summing the probability of each pixel until 90% is reached.
```

```python
#Sort the pixels by decending probability density
skymap.sort('PROBDENSITY', reverse=True)

#Area of each pixel
level, ipix = ah.uniq_to_level_ipix(skymap['UNIQ'])
pixel_area = ah.nside_to_pixel_area(ah.level_to_nside(level))

#Pixel area times the probability
prob = pixel_area * skymap['PROBDENSITY']

#Cummulative sum of probability
cumprob = np.cumsum(prob)

#Pixels for which cummulative is 0.9
i = cumprob.searchsorted(0.9)

#Sum of the areas of the pixels up to that one
area_90 = pixel_area[:i].sum()
area_90.to_value(u.deg**2)

> 

Reviewer: ```suggestion
```

```

> #### Flat Resolution HEALPix Sky maps

Reviewer: ```suggestion
### Flat Resolution HEALPix Sky maps
```

> Let's say you have a sky map fits file, read in Python with Healpy:

Reviewer: ```suggestion
Let's say you have a sky map fits file. To read in Python with Healpy:
```

Also, add some transition text to make it clear that you have switched here to flat-resolution files.

```python
import healpy as hp
import numpy as np

# Read both the HEALPix image data and the FITS header

> hpx, header = hp.read_map('*.fits.gz,0', h=True)

Reviewer: ```suggestion
hpx, header = hp.read_map('skymap.fits.gz', h=True)
```

> # Plotting a Mollweide-projection all-sky map

Reviewer: ```suggestion
# Plot a Mollweide-projection all-sky image
```

> np.mollview(hpx)

Reviewer: Also, if you actually want the plot to appear, then you have to add:

```python
from matplotlib import pyplot as plt

plt.show()
```

```

> Most Probable Sky Location

Reviewer: ```suggestion
#### Most Probable Sky Location
```

```python
# Find the highest probability pixel
ipix_max = np.argmax(hpx)

# Probability density per square degree at that position

> hpx[ipix_max] / hp.nside2pixarea(nside, degrees=True)

Reviewer: `nside` has not been defined.

Author: It's defined at `Probability Density at a Known Position` section

Reviewer: But that's from a different sky map.

# Highest probability pixel on the sky

> theta, phi = hp.pix2ang(nside, ipix_max)
> ra = np.rad2deg(phi)
> dec = np.rad2deg(0.5 * np.pi - theta)
> ra, dec

Reviewer: ```suggestion
ra, dec = hp.pix2ang(nside, ipix_max, lonlat=True)
ra, dec
```

```

#### Integrated probability in a Circle

We call hp.query_disc to obtain an array of indices for the pixels inside the circle.

```python

> # First define the Cartesian coordinates of the center of the Circle

Reviewer: ```suggestion
# First define the Cartesian coordinates of the center of the circle
```

ra = 180.0
dec = -45.0
radius = 2.5

> # We convert to spherical polar coordinates
> theta = 0.5 * np.pi - np.deg2rad(dec)
> phi = np.deg2rad(ra)
> radius = np.deg2rad(radius)
> 
> # Calculate Cartesian coordinates of the center of the Circle
> xyz = hp.ang2vec(theta, phi)

Reviewer: ```suggestion
# Calculate Cartesian coordinates of the center of the Circle
xyz = hp.ang2vec(ra, dec, lonlat=True)
```

# Obtain an array of indices for the pixels inside the circle

> ipix_disc = hp.query_disc(nside, xyz, radius)

Reviewer: ```suggestion
ipix_disc = hp.query_disc(nside, xyz, np.deg2rad(radius))
```

# Sum the probability in all of the matching pixels:
hpx[ipix_disc].sum()

```

#### Integrated probability in a Polygon

We can use the `hp.query_polygon` function to find the pixels inside a polygon and compute the probability that the source is inside the polygon by adding up the pixel values.

```python
#  Indices of the pixels within a polygon (defined by the Cartesian coordinates of its vertices)

xyz = [[-0.5, -0.4, -0.5],
        [-0.4, -0.4, -0.6],
        [-0.6, -0.3, -0.6],

>         [-0.7 , -0.4, -0.5]]

Reviewer: These are not unit vectors.

#### Other Documentation and HEALPix Packages

- Additional information can be found on the [LIGO website](https://emfollow.docs.ligo.org/userguide/tutorial/multiorder_skymaps.html)

- [healpy](https://healpy.readthedocs.io/en/latest/): Official python library for handling the pixlated data on sphere

- [astropy-healpix](https://pypi.org/project/astropy-healpix/): Integrates HEALPix with Astropy for data manipulation and analysis

- [mhealpy](https://mhealpy.readthedocs.io/en/latest/): Object-oriented wrapper of healpy for handling the multi-resolution maps

> - [MOCpy](https://cds-astro.github.io/mocpy/): Python library allowing easy creation, parsing and manipulation of Multi-Order Coverage maps.

Reviewer: If you want to present this as a bulleted list of links, then please start the section with a short paragraph (possibly just a single sentence) which is followed by the list.
